### PR TITLE
rustdoc: redesign toolbar and disclosure widgets

### LINKED ISSUE #77899: rustdoc: help widget oversized on mobile and can't scroll
Steps to reproduce:

1. Visit a documentation page on crates.io on a mobile browser (Brave 1.15.73 / Chromium 86, Android 11 on a Pixel 3).
2. Click the "?" icon next to the search box.

Expected result: Help dialog pops up

Actual result: Help dialog pops up, but is too big for the screen and not scrollable (see screenshot).

Note also that the keyboard shortcut section can probably be elided on mobile since there is usually no keyboard (though there can be!).

Screenshot_20201012-230926

### LINKED ISSUE #90310: Make the "help" button on generated rustdoc pages easier to find
I've talked to 3+ experienced rust developers this morning, all of whom were unaware of the existence of the help popup. They mentioned the following issues:
- The icon is to the right of search, not the left, making it seem unimportant.
- The icon is smaller than the search bar, making it seem unrelated.
- They've been trained by microsoft products that `?` is useless and doesn't have any non-trivial amount of info.

We should move the icon to the left and make it larger. They also suggested duplicating the help text on the docs.rs homepage, which seems like a much bigger hammer, but maybe is worth considering.

image

cc @rust-lang/rustdoc

### PR DESCRIPTION
**Pull request by notriddle**
Fixes #77899
Fixes #90310

## Preview

| before | after 
| ------ | -----
| image | image
| image | image
| image | image
| image | image
| image | image
| N/A | image
| image | image
| image | image

https://notriddle.com/rustdoc-html-demo-12/toolbar-v2/std/index.html

## Description

This adds labels to the icons and moves them away from the search box.

These changes are made together, because they work together, but are based on several complaints:

* The [+/-] thing are a Reddit-ism. They don't look like buttons, but look like syntax <https://rust-lang.zulipchat.com/#narrow/stream/266220-t-rustdoc/topic/More.20visual.20difference.20for.20the.20.2B.2F-.20.20Icons>, <https://github.com/rust-lang/rust/issues/59851> (some of these are laundry lists with more suggestions, but they all mention [+/-] looking wrong)

* The settings, help, and summary buttons are also too hard to recognize <https://lwn.net/Articles/987070/>, <https://github.com/rust-lang/rust/issues/90310>, <https://github.com/rust-lang/rust/issues/14475#issuecomment-274241997>, <https://internals.rust-lang.org/t/improve-rustdoc-design/12758> ("Not all functionality is self-explanatory, for example the [+] button in the top right corner, the theme picker or the settings button.")

The toggle-all and toggle-individual buttons both need done at once, since we want them to look like they go together. This changes them from both being [+/-] to both being arrows.

CC <https://github.com/rust-lang/rust/pull/113074#issuecomment-1677469680> and @jsha regarding the use of triangles for disclosure, which is what everyone wanted, but was pending a good toggle-all button. This PR adds a toggle-all button that should work.

Settings and Help are also migrated, so that the whole group can benefit from being described using actual words.

The breadcrumbs also get redesigned, so that they use less space, by shrinking the parent module path parts. This is done at the same time as the toolbar redesign because it's, effectively, moving space from the toolbar to the breadcrumbs.
This is aimed at avoiding any line wrapping at desktop sizes.

## Prior art

This style of toolbar, with explicit labels on the buttons, used to be more popular. It's not very common in web browsers nowadays, and for truly universal icons like :arrow_left: I can understand why, but words are great when icons fail.

image

### CODE REVIEW

**Comment by bjorn3** [comment]
These buttons are way too big on mobile IMO. Here is what it looks on a Galaxy S20 emulated by Firefox with the crate name replaced by a slightly larger one:

image

**Comment by bjorn3** [comment]
Much better!

**Comment by GuillaumeGomez** [comment]
Just to say: I LOVE IT! It looks so good. Well done!

**Comment by GuillaumeGomez** [comment]
So about the help button: it has important information like a link to the rustdoc book or the version of rustdoc used to generate this doc. So displaying it only on the search results page is not an improvement imo.

**Comment by GuillaumeGomez** [comment]
Maybe the problem here is that this menu contains too many different kind of information:
1. Search shortcuts
2. Search tricks
3. Link to rustdoc book
4. Version used to generate this doc

1 and 2 can be hidden away on mobile all the time, 3 and 4 seems like we're not providing all information we should.

**Comment by GuillaumeGomez** [comment]
> I'm not hiding the button because it contains too much information after you open it. I'm hiding the button because the button itself takes up space.

Yes. And my point was that hiding the help button, even as we do currently on mobile, is a bad idea as it contains many useful information. So maybe we should try to figure out what to do with this menu or eventually, how to not need these information to be part of this menu.

As for the buttons, maybe they're just too big? Why did you pick this size?

**Comment by GuillaumeGomez** [comment]
I thought about reducing both icon's and text's size and also reduce margins/paddings. But beyond that, I don't have any idea.

**Comment by GuillaumeGomez** [comment]
Sure. I'm in holiday currently so it might take some time. ^^'

**Comment by GuillaumeGomez** [comment]
I was thinking that we could reduce the width of the icons as the width of the viewport goes down. For example on iPad it would be:

Screenshot from 2024-09-02 16-24-52

To do this, I made the following changes:
 * reduce the icons width and height to 16px
 * reduce font size to 0.9em
 * Changed width of `#settings-menu > a, #help-button > a, button#toggle-all-docs` to `fit-content` and added `padding: 2px`.

I think the result is ok with this? And in any case, we remove the text under a given width so I think it's fine to also reduce size of icons as long as they're visible.

**Comment by GuillaumeGomez** [comment]
For this button, we can force its width to prevent that. As for the text, unlike other buttons in rustdoc, these ones both have text and an image, so I think it's fine if the text is a bit smaller.

**Comment by GuillaumeGomez** [comment]
I love it. :+1:

**Comment by GuillaumeGomez** [comment]
Should we hide the "show all" button on the search results?

**Comment by GuillaumeGomez** [comment]
I'm used to seeing disabled elements as "greyed out". Couldn't we make their color black (or the theme equivalent) and then grey it out?

**Comment by GuillaumeGomez** [comment]
Yep looks good! 👍

**Comment by GuillaumeGomez** [comment]
Looks good overall, just two things: the hover border is too close to the button's text and the border radius isn't big enough (for me).

So currently it looks like this:

Screenshot from 2024-09-03 21-31-56

With `padding-bottom: 4px` and `border-radius: 4px`:

image

What do you think?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Looks all good to me now.

@rfcbot poll

**Comment by camelid** [comment]
I'm wondering if it would be better to skip the trailing `::` on the parent module path. I like the idea of it in theory, to indicate that the "Struct `Vec`" part is attached, but it looks visually wrong, like there's something missing.

**Comment by lolbinarycat** [comment]
big fan of labeled icons that collapse to just icons on mobile.  docs.rs uses the same approach for its menubar, so there's some nice parity with that too.

**Comment by lolbinarycat** [comment]
one slight issue: the "copy path" button is probably a bit more confusing now that the module and name have been split up.

**Comment by camelid** [comment]
This looks amazing, thanks for the great work @notriddle. I only have one comment, and I don't feel strongly about it: What if instead of disabling "Summary" when in search view (etc.), we just delete it? We would also have to move it to be the leftmost icon, but I think that might make more sense anyway. It's a little strange to have content-related buttons after settings.

**Comment by camelid** [comment]
I checked my box since, like I said, I don't feel strongly, and it can always be tweaked later.

**Comment by lolbinarycat** [comment]
strictly speaking, the "collapse all" view would be more accurately described as an *outline*, not a summary, although it doesn't matter that much as long as it is clearly a button, since it's function will become apparent once it is clicked (just looking at the screenshots, it wasn't immediately obvious to me what it does)

**Comment by GuillaumeGomez** [comment]
Seems like we got the frontend team to agree with the changes. Thanks everyone! Time to approve it. Can't wait for this change to be available. :D

@bors r+ rollup

File: library/std/src/lib.rs [hunk 1]
```diff
@@ -32,13 +32,17 @@
...
 //! begin to find the verbosity of the prose distracting. At this stage in your
-//! development you may want to press the `[-]` button near the top of the
-//! page to collapse it into a more skimmable view.
-//!
-//! While you are looking at that `[-]` button also notice the `source`
-//! link. Rust's API documentation comes with the source code and you are
-//! encouraged to read it. The standard library source is generally high
-//! quality and a peek behind the curtains is often enlightening.
+//! development you may want to press the <code>
+//! <svg width="0.75rem" height="0.75rem" viewBox="0 0 12 12"
```
Comment (GuillaumeGomez) [inline library/std/src/lib.rs:36]: This failed in https://github.com/rust-lang/rust/pull/130701.

**Pull request merged**